### PR TITLE
Cera

### PR DESCRIPTION
**Pull request by C3ra906**
Hi Alex and Ariel, I'm having problem connecting the front-end with the-back end and I think one of the reasons might be that I'm trying to control the board with status messages we get with socket.on("move_done", ...) from the Board.js but, correct me if I'm wrong, it looks like the move function in the Board.js need to trigger socket.emit('move' ,...) in order for the server to get into socket.on("move", ...) to trigger socket.emit ('move_done' , ...)? So if we don't trigger move function at all in Board.js we can't get correct move_done status message from the server, is this right? You can try running the code I have for now but the status message is not matching up with what's happening on screen when I click. I'm going to take a long break and then try to work this out again but I'm wondering if we should not use socket for controlling how to mark the board or make the back-end check if a move is allowed outside of socket.on("move", ...).

### CODE REVIEW

**Comment by sashetov** [comment]
Yes Cera, you would need to do a
```
socket.emit("move", {"move_id": move_id})
```
where move_id is the slot on the board to move to

The status you get back should give you a clue if the move completed or not.
It wont return a good status if its not your turn, if the spot is taken etc.
If it completed you can allow the board state to update, if not you can show a dialog with an error message.

**Comment by gleason9113** [comment]
Alex:  Tell me if I'm getting the sequence of events wrong:
User clicks a box -> this calls the move function in  Board -> the Board function updates the boardState with the new move (validity checking happens here) then calls socket.emit('move', { boardState: newBoard }); -> the server fields the call to move, logs the id and data rec'd, then performs some checks.  Finally, if the checks pass, the server performs the move, checks for win and tie states, then emits the game data back to the client, right?

Assuming that's accurate, I think the server code can be simplified a bit.  The "is this move valid" checks happen clientside- the Board component won't call the move function on a 2nd click.  So on the server side, the move function can simply take the game data, update the board state with the new board state, check for winner/tie, then broadcast the updated game back to both players, right?

Frontend, then, we need to figure out how to receive that response and rerender the Board component with the new state.  Also, if I'm interpreting the output of the server-side move event, the "data" element is everything that came from the client, right?  Right now that just seems to be the board array itself- should the client be passing more stuff back?  Adding more fields to the boardstate shouldn't be too hard.

Of course, if I've misunderstood how the pieces are fitting together, this is all probably irrelevant, except for the point about the move-checking.

**Comment by C3ra906** [comment]
> Yes Cera, you would need to do a
> 
> ```
> socket.emit("move", {"move_id": move_id})
> ```
> 
> where move_id is the slot on the board to move to
> 
> The status you get back should give you a clue if the move completed or not. It wont return a good status if its not your turn, if the spot is taken etc. If it completed you can allow the board state to update, if not you can show a dialog with an error message.

Ohh, so I still test with the devtool. Thank you! I thought because I connected at least that portion of front end and back end we didn't need to do it within the devtool console...but that makes sense since we had to use devtool console to add players. Will test it out later today.

**Comment by sashetov** [comment]
> User clicks a box -> this calls the move function in Board -> the Board function updates the boardState with the new move (validity checking happens here) then calls socket.emit('move', { boardState: newBoard }); -> the server fields the call to move, logs the id and data rec'd, then performs some checks. Finally, if the checks pass, the server performs the move, checks for win and tie states, then emits the game data back to the client, right?
yes, sounds right


> Assuming that's accurate, I think the server code can be simplified a bit. The "is this move valid" checks happen clientside- the Board component won't call the move function on a 2nd click. So on the server side, the move function can simply take the game data, update the board state with the new board state, check for winner/tie, then broadcast the updated game back to both players, right?

the server will always check for a bad move because it has to preserve the board state, it cannot trust input from the client.
a malicious actor in the client could just send whatever it wants with some javascript and mess up the server state
so, no, i don't think we can simplify it much more, no

> Frontend, then, we need to figure out how to receive that response and rerender the Board component with the new state. Also, if I'm interpreting the output of the server-side move event, the "data" element is everything that came from the client, right? Right now that just seems to be the board array itself- should the client be passing more stuff back? Adding more fields to the boardstate shouldn't be too hard.

well, you already have socket handlers, you just need to read the error messages and status "ok" or "error" and craft your logic around that.

if you notice bugs in the backend let me know, its possible i have some, though i didn't notice any when testing manually

> Of course, if I've misunderstood how the pieces are fitting together, this is all probably irrelevant, except for the point about the move-checking.

from my standpoint, what needs to happen is:

1. /join call to join game
2. before you can start playing another player needs to join, so display a loading screen and don't show the board or something similar
2.1. you can check whether other player has joined by emitting isOpponentAvailable on the socket and handling the opponentAvailable return message with a handler
2.1.1 you may have to emit isOpponentAvailable every few seconds or something like that until the handler for opponentAvailable fires, at which point you can stop that loop
3. once you know they have joined, the event data in isOpponentAvailable will tell you if its your turn or not, but you can also check that with the data you got in the /join response, because if your symbol is 'O' there, its your turn (since that's the symbol for people that joined first)
3.1. if its not your turn you cannot move and if you attempt to move it should show some warning
3.2. if you decide to hack it and send a socket emit to make it move and its not your turn it will fail, so if you are looking at errors coming out of the socket and building it around that logic, you can handle it that way as well
3.2.1. if it is your turn you attempt to make a turn. if you try to move to a spot that is already taken it will fail, if the game is over it will fail with a different message. check these error messages, build your logic around them
4. after move_done comes back, if status is ok, you will be told not only the board state, which you should sync from the backend to the frontend but also whether there is a winner or not.
4.1. if the game is over it should change the component to some kind of screen indicating that there was a win, loss or draw
    -1 means error because game doesn't have both players yet <-- you should never get this
    0 means player 1 won
    1 means player 2 won
    2 means game is over and was a draw
4.1.1 player 1 and 2 may be unknown to you, i can change that so that it tells you which socket id won, instead

**Comment by sashetov** [comment]
@gleason9113 lets talk tomorrow if you have time it will be easier to explain, i can share my screen etc

File: src/components/Board.js [hunk 1]
```diff
@@ -24,125 +29,171 @@ const Board = ({ socket }) => {
...
+    }
+  };
 
   const move = (event, index) => {
     const newBoard = [...boardState];
     const curr = turn === 0 ? 'X' : 'O';
-    if (newBoard[index] === '') {
-      newBoard[index] = curr;
-      event.target.innerHTML = curr;
+    socket.on('opponentAvailable', two_player); //status ok
```
Comment (sashetov) [inline src/components/Board.js:62]: this part seems to work and the board appears after second person shows up.

**Review by sashetov — COMMENTED** [review]
looks good but I did notice some issue which I'm not sure is server side or not, where you can't move at all, regardless of which player

i think there is a bug, i can't tell yet if server side or not but you can't move anywhere.

ill debug it a bit to determine where it is and then say if it can be approved or not

**Comment by gleason9113** [comment]
Ran it and didn't encounter an issue there - the moves are appearing client and server side for me.  Everything else looks good, as far as I can tell.  Approved, unless Alex finds his bug.

**Comment by sashetov** [comment]
i think this pr should work after we merge in #11 but I'd want to test just to be sure

**Comment by C3ra906** [comment]
I'm going to close this one and work off of the newly merged main!

**Pull request closed without merging**